src/uct/tcp/tcp_iface.c

```diff
@@ -615,6 +617,85 @@ static UCS_CLASS_DEFINE_NEW_FUNC(uct_tcp_iface_t, uct_iface_t, uct_md_h,
                                  uct_worker_h, const uct_iface_params_t*,
                                  const uct_iface_config_t*);
 
+/* Fetch information about available network devices through an ioctl.  */
+static ucs_status_t uct_tcp_query_devices_ioctl(uct_md_h md,
+                                                uct_tl_device_resource_t **devices_p,
+                                                unsigned *num_devices_p)
+{
+    int sock, err, i;
+    uct_tl_device_resource_t *devices, *tmp;
+    unsigned num_devices;
+    ucs_status_t status;
+    struct ifconf conf;
+
+    conf.ifc_len = 0;
+    conf.ifc_req = NULL;
+
+    status = ucs_socket_create(AF_INET, SOCK_STREAM, &sock);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    err = ioctl(sock, SIOCGIFCONF, &conf);
+    if (err < 0) {
+        ucs_error("ioctl(SIOCGIFCONF) failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto out;
+    }
+
+    conf.ifc_req = ucs_malloc(1, conf.ifc_len, "ifreq");
+    if (conf.ifc_req == NULL) {
+        ucs_error("memory alocation failed");
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    err = ioctl(sock, SIOCGIFCONF, &conf);
+    if (err < 0) {
+        ucs_error("ioctl(SIOCGIFCONF) failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto out_free;
+    }
+
+    devices     = NULL;
+    num_devices = 0;
+    for (i = 0; i < (conf.ifc_len / sizeof(struct ifreq)); i++) {
+        const char *name = conf.ifc_req[i].ifr_name;
+
+        if (!ucs_netif_is_active(name)) {
+            continue;
+        }
+
+        tmp = ucs_realloc(devices, sizeof(*devices) * (num_devices + 1),
+                          "tcp devices");
+        if (tmp == NULL) {
+            ucs_free(devices);
+            status = UCS_ERR_NO_MEMORY;
+            goto out_free;
+        }
+        devices = tmp;
+
+        ucs_snprintf_zero(devices[num_devices].name,
+                          sizeof(devices[num_devices].name),
+                          "%s", name);
+        devices[num_devices].type = UCT_DEVICE_TYPE_NET;
+        ++num_devices;
+    }
+
+    *num_devices_p = num_devices;
+    *devices_p     = devices;
+    status         = UCS_OK;
+
+out_free:
+    ucs_free(conf.ifc_req);
+
```

Comment:
could you remove the blank line (691) pls?

Reply:
Sure, done.